statime-linux/src/clock/raw.rs

```diff
@@ -71,86 +107,96 @@ impl RawLinuxClock {
     ///
     /// If the time offset is higher than 0.5 seconds, then the clock will be
     /// set directly and no frequency change will be made.
-    pub fn adjust_clock(&mut self, time_offset: f64, frequency_multiplier: f64) -> Result<(), i32> {
+    pub fn adjust_clock(
+        &mut self,
+        time_offset: f64,
+        frequency_multiplier: f64,
+    ) -> Result<(), Error> {
         log::trace!("Adjusting clock: {time_offset}s, {frequency_multiplier}x");
 
-        let (current_timex, _clock_state) = self.get_clock_state()?;
+        self.adjust_frequency(frequency_multiplier)?;
+        self.adjust_offset(time_offset)
+    }
 
-        // We do an offset with precision
-        let mut frequency_timex = Timex::new();
-
-        frequency_timex.set_status(
-            current_timex.get_status()
-                | StatusFlags::FREQHOLD // We want no automatic frequency updates
-                & !StatusFlags::PLL
-                & !StatusFlags::PPSFREQ
-                & !StatusFlags::FLL
-                & !StatusFlags::PPSTIME,
-        );
-
-        frequency_timex.set_mode(
-            AdjustFlags::FREQUENCY, // We'll be setting the frequency as well
-        );
-
-        // We need to change the ppm value to a speed factor so we can use
-        // multiplication to get the new frequency
-        let current_ppm = current_timex.get_frequency();
+    fn adjust_frequency(&mut self, frequency_multiplier: f64) -> Result<(), Error> {
+        let mut timex = EMPTY_TIMEX;
+        self.adjtime(&mut timex)?;
+
+        const M: f64 = 1_000_000.0;
+
+        // In struct timex, freq, ppsfreq, and stabil are ppm (parts per million) with a
+        // 16-bit fractional part, which means that a value of 1 in one of those fields
+        // actually means 2^-16 ppm, and 2^16=65536 is 1 ppm.  This is the case for both
+        // input values (in the case of freq) and output values.
+        let current_ppm = (timex.freq >> 16) as f64 + ((timex.freq & 0xffff) as f64 / 65536.0);
```

Comment:
This can be simplified:
```suggestion
        let current_ppm = timex.freq as f64 / 65536.0;
```